### PR TITLE
SQL-on-FHIR processResource

### PR DESCRIPTION
**Pull request by codyebberson**
I got nerd sniped by @jdjkelly's discussion: https://github.com/medplum/medplum/discussions/4661

This PR implements the main skeleton of the SQL-on-FHIR `processResource` and `process` algorithms.

See: https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/implementer_guidance.html#process-a-resource-entry-point

With this, we're at 57% pass rate

```
$ npm t -- src/sql-on-fhir/ --verbose=false

Test Suites: 1 failed, 1 total
Tests:       50 failed, 67 passed, 117 total
Snapshots:   0 total
Time:        1.274 s
Ran all test suites matching /src\\sql-on-fhir\\/i.
npm ERR! Lifecycle script `test` failed with error:
npm ERR! Error: command failed
npm ERR!   in workspace: @medplum/core@3.1.8
npm ERR!   at location: C:\Users\cody\dev\medplum\packages\core
```

Examples of open issues:
* Some new required FHIRPath functions: https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/StructureDefinition-ViewDefinition.html#required-additional-functions
* Long tail error handling (i.e., column order mismatch in unions)
* Some cases of existing FHIRPath that we do not implement (i.e., default separator for `join()`)

Questions:
* Is this ok in `@medplum/core`, or should it be a separate package (i.e., `@medplum/sql-on-fhir`)?
* Ok to copy the test files direct?  The implementation guide recommends git submodules, but git submodules.
* Rather than treating as pass/fail, I plan to temporarily mark this as "passing" to generate the test report.  That could get us on the board: https://fhir.github.io/sql-on-fhir-v2/#impls

### CODE REVIEW

File: packages/core/src/sql-on-fhir/eval.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
+import { Resource, ViewDefinition, ViewDefinitionSelect } from '@medplum/fhirtypes';
```
Comment (jdjkelly) [inline packages/core/src/sql-on-fhir/eval.ts:1]: I got these types working on a branch locally too. Curious how you approached it (it's not committed in the PR). I loaded up the structure definitions manually into @medplum/generator from the IG's package.

This is actually something that's generally _very_ useful and under-explored in TypeScript FHIR support - typings for profiles.

**Comment by jdjkelly** [comment]
This is pretty promising...

> Ok to copy the test files direct? The implementation guide recommends git submodules, but git submodules.

This probably makes sense - this is basically the point of this shared test suite. Take the tests, implement a function that generates a report on the test cases that can be committed back to the project to show community implementation progress.

@Diferti could help chase down test coverage here - he handled some corner cases for the reference implementation (lowBoundary and highBoundary support)

File: packages/core/src/sql-on-fhir/eval.ts (hunk 1 above)
Comment (codyebberson) [reply packages/core/src/sql-on-fhir/eval.ts:1]: Yes, sorry for confusion, the types were added in separate PR: https://github.com/medplum/medplum/pull/4674

For lack of a better place, we stuffed them in `packages/definitions/dist/fhir/r4/profiles-medplum.json` for now.

**Comment by codyebberson** [comment]
Current errors:

```json
[
  "basic.json: where returns non-boolean for some cases: Error: WHERE clause must evaluate to a boolean",
  "collection.json: collection = false relative to forEach parent: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoMatchObject\u001b[2m(\u001b[22m\u001b[32mexpected\u001b[39m\u001b[2m)\u001b[22m\n\n\u001b[32m- Expected  - 6\u001b[39m\n\u001b[31m+ Received  + 2\u001b[39m\n\n\u001b[33m@@ -1,10 +1,8 @@\u001b[39m\n\u001b[2m  Array [\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"first_name\": Array [\u001b[39m\n\u001b[32m-       \"g1.1\",\u001b[39m\n\u001b[32m-     ],\u001b[39m\n\u001b[31m+     \"first_name\": \"g1.1\",\u001b[39m\n\u001b[2m      \"id\": \"pt1\",\u001b[22m\n\u001b[2m      \"last_name\": \"f1.1\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[2m      \"first_name\": Array [\u001b[22m\n\u001b[33m@@ -13,13 +11,11 @@\u001b[39m\n\u001b[2m      ],\u001b[22m\n\u001b[2m      \"id\": \"pt1\",\u001b[22m\n\u001b[2m      \"last_name\": \"f1.2\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"first_name\": Array [\u001b[39m\n\u001b[32m-       \"g2.1\",\u001b[39m\n\u001b[32m-     ],\u001b[39m\n\u001b[31m+     \"first_name\": \"g2.1\",\u001b[39m\n\u001b[2m      \"id\": \"pt2\",\u001b[22m\n\u001b[2m      \"last_name\": \"f2.1\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[2m      \"first_name\": Array [\u001b[22m",
  "collection.json: collection = false relative to forEachOrNull parent: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoMatchObject\u001b[2m(\u001b[22m\u001b[32mexpected\u001b[39m\u001b[2m)\u001b[22m\n\n\u001b[32m- Expected  - 6\u001b[39m\n\u001b[31m+ Received  + 2\u001b[39m\n\n\u001b[33m@@ -1,10 +1,8 @@\u001b[39m\n\u001b[2m  Array [\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"first_name\": Array [\u001b[39m\n\u001b[32m-       \"g1.1\",\u001b[39m\n\u001b[32m-     ],\u001b[39m\n\u001b[31m+     \"first_name\": \"g1.1\",\u001b[39m\n\u001b[2m      \"id\": \"pt1\",\u001b[22m\n\u001b[2m      \"last_name\": \"f1.1\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[2m      \"first_name\": Array [\u001b[22m\n\u001b[33m@@ -13,13 +11,11 @@\u001b[39m\n\u001b[2m      ],\u001b[22m\n\u001b[2m      \"id\": \"pt1\",\u001b[22m\n\u001b[2m      \"last_name\": \"f1.2\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"first_name\": Array [\u001b[39m\n\u001b[32m-       \"g2.1\",\u001b[39m\n\u001b[32m-     ],\u001b[39m\n\u001b[31m+     \"first_name\": \"g2.1\",\u001b[39m\n\u001b[2m      \"id\": \"pt2\",\u001b[22m\n\u001b[2m      \"last_name\": \"f2.1\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[2m      \"first_name\": Array [\u001b[22m",
  "constant.json: constant in path: Error: FhirPathError on \"name.where(use = %name_use).family\": Error: Undefined variable %name_use",
  "constant.json: constant in forEach: Error: FhirPathError on \"name.where(use = %name_use)\": Error: Undefined variable %name_use",
  "constant.json: constant in where element: Error: FhirPathError on \"name.where(use = %name_use).exists()\": Error: Undefined variable %name_use",
  "constant.json: constant in unionAll: Error: FhirPathError on \"name.where(use = %use1)\": Error: Undefined variable %use1",
  "constant.json: integer constant: Error: FhirPathError on \"name[%name_index].family\": Error: Undefined variable %name_index",
  "constant.json: boolean constant: Error: FhirPathError on \"deceased.ofType(boolean).exists() and deceased.ofType(boolean) = %is_deceased\": Error: Undefined variable %is_deceased",
  "constant_types.json: base64Binary: Error: FhirPathError on \"udiCarrier.first().carrierAIDC = %aidc\": Error: Undefined variable %aidc",
  "constant_types.json: code: Error: FhirPathError on \"gender = %gender\": Error: Undefined variable %gender",
  "constant_types.json: date: Error: FhirPathError on \"birthDate = %bd\": Error: Undefined variable %bd",
  "constant_types.json: dateTime: Error: FhirPathError on \"identified.ofType(DateTime) = %id_time\": Error: Undefined variable %id_time",
  "constant_types.json: decimal: Error: FhirPathError on \"value.ofType(Quantity).value < %v\": Error: Undefined variable %v",
  "constant_types.json: id: Error: FhirPathError on \"output.first().value.ofType(id) = %id\": Error: Undefined variable %id",
  "constant_types.json: instant: Error: FhirPathError on \"effective.ofType(Instant) = %eff\": Error: Undefined variable %eff",
  "constant_types.json: oid: Error: FhirPathError on \"output.first().value.ofType(oid) = %oid\": Error: Undefined variable %oid",
  "constant_types.json: positiveInt: Error: FhirPathError on \"item.first().itemSequence = %seq\": Error: Undefined variable %seq",
  "constant_types.json: time: Error: FhirPathError on \"value.ofType(Time) = %t\": Error: Undefined variable %t",
  "constant_types.json: unsignedInt: Error: FhirPathError on \"numberOfSeries = %series\": Error: Undefined variable %series",
  "constant_types.json: uri: Error: FhirPathError on \"url = %uri\": Error: Undefined variable %uri",
  "constant_types.json: url: Error: FhirPathError on \"output.first().value.ofType(url) = %url\": Error: Undefined variable %url",
  "constant_types.json: uuid: Error: FhirPathError on \"output.first().value.ofType(uuid) = %uuid\": Error: Undefined variable %uuid",
  "fhirpath.json: collection: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoMatchObject\u001b[2m(\u001b[22m\u001b[32mexpected\u001b[39m\u001b[2m)\u001b[22m\n\n\u001b[32m- Expected  - 1\u001b[39m\n\u001b[31m+ Received  + 1\u001b[39m\n\n\u001b[33m@@ -10,8 +10,8 @@\u001b[39m\n\u001b[2m        \"f2.1\",\u001b[22m\n\u001b[2m        \"f2.2\",\u001b[22m\n\u001b[2m      ],\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"v\": Array [],\u001b[39m\n\u001b[31m+     \"v\": null,\u001b[39m\n\u001b[2m    },\u001b[22m\n\u001b[2m  ]\u001b[22m",
  "fhirpath.json: string join: Error: FhirPathError on \"name.given.join(', ' )\": Error: Unrecognized function: join",
  "fhirpath.json: string join: default separator: Error: FhirPathError on \"name.given.join()\": Error: Unrecognized function: join",
  "fhirpath_numbers.json: add observation: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoMatchObject\u001b[2m(\u001b[22m\u001b[32mexpected\u001b[39m\u001b[2m)\u001b[22m\n\n\u001b[32m- Expected  - 9\u001b[39m\n\u001b[31m+ Received  + 9\u001b[39m\n\n\u001b[2m  Array [\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"add\": 5,\u001b[39m\n\u001b[32m-     \"div\": 1.5,\u001b[39m\n\u001b[32m-     \"eq\": false,\u001b[39m\n\u001b[32m-     \"ge\": true,\u001b[39m\n\u001b[32m-     \"gt\": true,\u001b[39m\n\u001b[31m+     \"add\": null,\u001b[39m\n\u001b[31m+     \"div\": null,\u001b[39m\n\u001b[31m+     \"eq\": null,\u001b[39m\n\u001b[31m+     \"ge\": null,\u001b[39m\n\u001b[31m+     \"gt\": null,\u001b[39m\n\u001b[2m      \"id\": \"o1\",\u001b[22m\n\u001b[32m-     \"le\": false,\u001b[39m\n\u001b[32m-     \"lt\": false,\u001b[39m\n\u001b[32m-     \"mul\": 6,\u001b[39m\n\u001b[32m-     \"sub\": 1,\u001b[39m\n\u001b[31m+     \"le\": null,\u001b[39m\n\u001b[31m+     \"lt\": null,\u001b[39m\n\u001b[31m+     \"mul\": null,\u001b[39m\n\u001b[31m+     \"sub\": null,\u001b[39m\n\u001b[2m    },\u001b[22m\n\u001b[2m  ]\u001b[22m",
  "fn_boundary.json: decimal lowBoundary: Error: FhirPathError on \"value.ofType(Quantity).value.lowBoundary()\": Error: Unrecognized function: lowBoundary",
  "fn_boundary.json: decimal highBoundary: Error: FhirPathError on \"value.ofType(Quantity).value.highBoundary()\": Error: Unrecognized function: highBoundary",
  "fn_boundary.json: datetime lowBoundary: Error: FhirPathError on \"value.ofType(dateTime).lowBoundary()\": Error: Unrecognized function: lowBoundary",
  "fn_boundary.json: datetime highBoundary: Error: FhirPathError on \"value.ofType(dateTime).highBoundary()\": Error: Unrecognized function: highBoundary",
  "fn_boundary.json: date lowBoundary: Error: FhirPathError on \"birthDate.lowBoundary()\": Error: Unrecognized function: lowBoundary",
  "fn_boundary.json: date highBoundary: Error: FhirPathError on \"birthDate.highBoundary()\": Error: Unrecognized function: highBoundary",
  "fn_boundary.json: time lowBoundary: Error: FhirPathError on \"value.ofType(time).lowBoundary()\": Error: Unrecognized function: lowBoundary",
  "fn_boundary.json: time highBoundary: Error: FhirPathError on \"value.ofType(time).highBoundary()\": Error: Unrecognized function: highBoundary",
  "fn_extension.json: simple extension: Error: FhirPathError on \"extension('http://hl7.org/fhir/us/core/StructureDefinition/us-core-birthsex').value.ofType(code).first()\": Error: Unrecognized function: extension",
  "fn_extension.json: nested extension: Error: FhirPathError on \"extension('http://hl7.org/fhir/us/core/StructureDefinition/us-core-race').extension('ombCategory').value.ofType(Coding).code.first()\": Error: Unrecognized function: extension",
  "fn_join.json: join with comma: Error: FhirPathError on \"name.given.join(',')\": Error: Unrecognized function: join",
  "fn_join.json: join with empty value: Error: FhirPathError on \"name.given.join('')\": Error: Unrecognized function: join",
  "fn_join.json: join with no value - default to no separator: Error: FhirPathError on \"name.given.join()\": Error: Unrecognized function: join",
  "fn_reference_keys.json: getReferenceKey result matches getResourceKey without type specifier: Error: FhirPathError on \"getResourceKey() = link.other.getReferenceKey()\": Error: Unrecognized function: getResourceKey",
  "fn_reference_keys.json: getReferenceKey result matches getResourceKey with right type specifier: Error: FhirPathError on \"getResourceKey() = link.other.getReferenceKey(Patient)\": Error: Unrecognized function: getResourceKey",
  "fn_reference_keys.json: getReferenceKey result matches getResourceKey with wrong type specifier: Error: FhirPathError on \"link.other.getReferenceKey(Observation)\": Error: Unrecognized function: getReferenceKey",
  "foreach.json: forEach: two on the same level: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoMatchObject\u001b[2m(\u001b[22m\u001b[32mexpected\u001b[39m\u001b[2m)\u001b[22m\n\n\u001b[32m- Expected  - 4\u001b[39m\n\u001b[31m+ Received  + 4\u001b[39m\n\n\u001b[33m@@ -2,16 +2,16 @@\u001b[39m\n\u001b[2m    Object {\u001b[22m\n\u001b[2m      \"cont_family\": \"FC1.1\",\u001b[22m\n\u001b[2m      \"pat_family\": \"F1.1\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"cont_family\": \"FC1.2\",\u001b[39m\n\u001b[32m-     \"pat_family\": \"F1.1\",\u001b[39m\n\u001b[32m-   },\u001b[39m\n\u001b[32m-   Object {\u001b[39m\n\u001b[2m      \"cont_family\": \"FC1.1\",\u001b[22m\n\u001b[2m      \"pat_family\": \"F1.2\",\u001b[22m\n\u001b[31m+   },\u001b[39m\n\u001b[31m+   Object {\u001b[39m\n\u001b[31m+     \"cont_family\": \"FC1.2\",\u001b[39m\n\u001b[31m+     \"pat_family\": \"F1.1\",\u001b[39m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[2m      \"cont_family\": \"FC1.2\",\u001b[22m\n\u001b[2m      \"pat_family\": \"F1.2\",\u001b[22m\n\u001b[2m    },\u001b[22m",
  "foreach.json: forEachOrNull & unionAll on the same level: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoMatchObject\u001b[2m(\u001b[22m\u001b[32mexpected\u001b[39m\u001b[2m)\u001b[22m\n\n\u001b[32m- Expected  - 2\u001b[39m\n\u001b[31m+ Received  + 0\u001b[39m\n\n\u001b[33m@@ -19,12 +19,10 @@\u001b[39m\n\u001b[2m      \"id\": \"pt1\",\u001b[22m\n\u001b[2m      \"name\": \"N2\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[2m      \"id\": \"pt2\",\u001b[22m\n\u001b[32m-     \"name\": null,\u001b[39m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[2m      \"id\": \"pt3\",\u001b[22m\n\u001b[32m-     \"name\": null,\u001b[39m\n\u001b[2m    },\u001b[22m\n\u001b[2m  ]\u001b[22m",
  "foreach.json: forEachOrNull & unionAll & column & select on the same level: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoMatchObject\u001b[2m(\u001b[22m\u001b[32mexpected\u001b[39m\u001b[2m)\u001b[22m\n\n\u001b[32m- Expected  - 4\u001b[39m\n\u001b[31m+ Received  + 0\u001b[39m\n\n\u001b[33m@@ -28,17 +28,13 @@\u001b[39m\n\u001b[2m      \"id\": \"pt1\",\u001b[22m\n\u001b[2m      \"name\": \"N2\",\u001b[22m\n\u001b[2m      \"tel_system\": \"email\",\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"gender\": null,\u001b[39m\n\u001b[2m      \"id\": \"pt2\",\u001b[22m\n\u001b[32m-     \"name\": null,\u001b[39m\n\u001b[2m      \"tel_system\": null,\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m    Object {\u001b[22m\n\u001b[32m-     \"gender\": null,\u001b[39m\n\u001b[2m      \"id\": \"pt3\",\u001b[22m\n\u001b[32m-     \"name\": null,\u001b[39m\n\u001b[2m      \"tel_system\": null,\u001b[22m\n\u001b[2m    },\u001b[22m\n\u001b[2m  ]\u001b[22m",
  "logic.json: filtering with 'or': Error: WHERE clause must evaluate to a boolean",
  "union.json: column mismatch: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoThrow\u001b[2m()\u001b[22m\n\nReceived function did not throw",
  "union.json: column order mismatch: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoThrow\u001b[2m()\u001b[22m\n\nReceived function did not throw",
  "validate.json: wrong fhirpath: Error: \u001b[2mexpect(\u001b[22m\u001b[31mreceived\u001b[39m\u001b[2m).\u001b[22mtoThrow\u001b[2m()\u001b[22m\n\nReceived function did not throw"
]
```

File: packages/core/src/sql-on-fhir/eval.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+      // Append rows to parts
+      // (Note: do not append the elements but the whole list, so the final element of parts is now the list rows)
+      parts.push(rows);
+    }
+
+    // Process UnionAlls:
+    // Initialize urows as an empty list of rows
+    // For each selection structure u of S.unionAll
+    if (s.unionAll) {
```
Comment (mattlong) [inline packages/core/src/sql-on-fhir/eval.ts:154]: Should this be wrapped in the `if (s.unionAll)` check? Reading the pseudocode, it seems like you're still supposed to append an empty list if `S.unionAll` is missing. It may not matter given the cartesianProduct step below, but wanted to call it out.

**Review by mattlong — APPROVED** [review]
LGTM. Just one comment that may be a noop.

File: packages/core/src/sql-on-fhir/eval.ts (hunk 2 above)
Comment (codyebberson) [reply packages/core/src/sql-on-fhir/eval.ts:154]: Great question, and astute observation.

My first stab did exactly that; however, it doesn't work.

The reason is that it ends with a call to `parts.push(urows);`  If `urows` is empty, you then have an empty array in `parts`. When calculating the cartesian product, an empty array is like multiplying by zero, it wipes out everything.

It could be handled by doing `if (urows.length > 0) { parts.push(urows); }`, or updating the cartesian product logic, but this felt cleaner 🤷
